src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let referenceState = { ...state };
+
+  for (const action of actions) {
+    const copy = { ...referenceState };
+    const { type } = action;
+
+    switch (type) {
+      case 'addProperties': {
+        const { extraData } = action;
+
+        for (const [key, value] of Object.entries(extraData)) {
+          copy[key] = value;
+        }
+
+        stateHistory.push(copy);
+        referenceState = { ...copy };
+        break;
+      }
+
+      case 'removeProperties': {
+        const { keysToRemove } = action;
+
+        for (const key of keysToRemove) {
+          delete copy[key];
+        }
+
+        stateHistory.push(copy);
+        referenceState = { ...copy };
+        break;
+      }
+
+      case 'clear': {
+        stateHistory.push({});
+        referenceState = {};
```

Comment:
[CODE STYLE] - You are directly pushing an empty object to 'stateHistory' and setting 'referenceState' to an empty object within the 'clear' case. This is correct, but to maintain consistency with other cases and avoid direct mutation of 'stateHistory', consider using the same approach as in other cases.